cmd/promoted-image-governor/main.go

```diff
@@ -0,0 +1,314 @@
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/test-infra/prow/flagutil"
+	"k8s.io/test-infra/prow/interrupts"
+	"k8s.io/test-infra/prow/logrusutil"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	imagev1 "github.com/openshift/api/image/v1"
+
+	"github.com/openshift/ci-tools/pkg/api"
+	"github.com/openshift/ci-tools/pkg/config"
+	"github.com/openshift/ci-tools/pkg/steps/release"
+)
+
+type options struct {
+	ciOperatorconfigPath             string
+	kubeconfig                       string
+	dryRun                           bool
+	ignoredImageStreamTagsRaw        flagutil.Strings
+	ignoredImageStreamTags           []*regexp.Regexp
+	releaseControllerMirrorConfigDir string
+}
+
+func parseOptions() *options {
+	opts := &options{}
+	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+	// Controller-Runtimes root package imports the package that sets this flag
+	kubeconfigFlagDescription := "Path to the kubeconfig file to use for CLI requests."
+	if f := fs.Lookup("kubeconfig"); f != nil {
+		f.Usage = kubeconfigFlagDescription
+		// https://i.kym-cdn.com/entries/icons/original/000/018/012/this_is_fine.jpeg
+		defer func() { opts.kubeconfig = f.Value.String() }()
+	} else {
+		fs.StringVar(&opts.kubeconfig, "kubeconfig", "", kubeconfigFlagDescription)
+	}
+	fs.StringVar(&opts.ciOperatorconfigPath, "ci-operator-config-path", "", "Path to the ci operator config")
+	fs.BoolVar(&opts.dryRun, "dry-run", true, "Whether to run the controller-manager with dry-run")
+	fs.Var(&opts.ignoredImageStreamTagsRaw, "ignored-image-stream-tags", "A regex to match tag in the form of namespace/name:tag format. Can be passed multiple times.")
+	fs.StringVar(&opts.releaseControllerMirrorConfigDir, "release-controller-mirror-config-dir", "", "Path to the release controller mirror config directory")
+	if err := fs.Parse(os.Args[1:]); err != nil {
+		logrus.WithError(err).Fatal("could not parse input")
+	}
+	return opts
+}
+
+func (o *options) validate() error {
+	if o.ciOperatorconfigPath == "" {
+		return fmt.Errorf("--ci-operator-config-path must be set")
+	}
+	if o.releaseControllerMirrorConfigDir == "" {
+		return fmt.Errorf("--release-controller-mirror-config-dir must be set")
+	}
+	for _, s := range o.ignoredImageStreamTagsRaw.Strings() {
+		re, err := regexp.Compile(s)
+		if err != nil {
+			return fmt.Errorf("failed to compile regex from %q: %w", s, err)
+		}
+		o.ignoredImageStreamTags = append(o.ignoredImageStreamTags, re)
+	}
+	return nil
+}
+
+func addSchemes() error {
+	if err := imagev1.AddToScheme(scheme.Scheme); err != nil {
+		return fmt.Errorf("failed to add imagev1 to scheme: %w", err)
+	}
+	return nil
+}
+
+func validatePromotedTags(tags []api.ImageStreamTagReference) error {
```

Comment:
You could simplify this:

```go
tags := map[api.ImageStreamTagReference]interface{}{}
var duplicates []string
for _, tag := tags {
    if _, exists := tags[tag]; exists {
        duplicates = append(duplicates, tag.ISTagName())
    } else {
        tags[tag] = nil
    }
}
if len(duplicates) > 0 {
    return fmt.Errorf("found tags promoted more than once: %s", dups)
}
return nil

Reply:
However, we already do this validation in `ci-operator-checkconfig` and I don't know that we need to duplicate it here ...

Reply:
Removed.